### PR TITLE
[IOTDB-3714] Refactor operators related to aggregate query to support batch processing

### PR DESCRIPTION
**Pull request by liuminghui233**
details see [doc](https://apache-iotdb.feishu.cn/docs/doccnuuXYJo0AeM34YhCq61o6Ce#1GG9GP)

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java [hunk 1]
```diff
@@ -57,6 +59,60 @@ public AlignedSeriesScanUtil(
...
+      throw new IOException("Can only get statistics by index from alignedChunkMetaData");
+    }
+    return ((AlignedChunkMetadata) firstChunkMetadata).getStatistics(index);
+  }
+
+  @Override
+  protected Statistics currentChunkTimeStatistics() throws IOException {
+    if (!(firstChunkMetadata instanceof AlignedChunkMetadata)) {
+      throw new IOException("Can only get statistics of time column from alignedChunkMetaData");
+    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:90]: same as above.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java [hunk 2]
```diff
@@ -57,6 +59,60 @@ public AlignedSeriesScanUtil(
...
+  }
+
+  @Override
+  protected Statistics currentPageStatistics(int index) throws IOException {
+    if (firstPageReader == null) {
+      return null;
+    }
+    if (!(firstPageReader.isAlignedPageReader())) {
+      throw new IOException("Can only get statistics by index from AlignedPageReader");
+    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:101]: same as above.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java [hunk 3]
```diff
@@ -57,6 +59,60 @@ public AlignedSeriesScanUtil(
...
+  }
+
+  @Override
+  protected Statistics currentPageTimeStatistics() throws IOException {
+    if (firstPageReader == null) {
+      return null;
+    }
+    if (!(firstPageReader.isAlignedPageReader())) {
+      throw new IOException("Can only get statistics of time column from AlignedPageReader");
+    }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:112]: same as above.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/SeriesAggregationScanOperator.java [hunk 4]
```diff
@@ -96,337 +57,6 @@ public SeriesAggregationScanOperator(
...
-    return !seriesScanUtil.isPageOverlapped()
-        && containedByTimeFilter(currentPageStatistics)
-        && !seriesScanUtil.currentPageModified();
-  }
-
-  private boolean containedByTimeFilter(Statistics statistics) {
-    Filter timeFilter = seriesScanUtil.getTimeFilter();
-    return timeFilter == null
-        || timeFilter.containStartEndTime(statistics.getStartTime(), statistics.getEndTime());
+    this.subSensorSize = 1;
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/SeriesAggregationScanOperator.java:60]: put this field into super's constructor.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/SeriesAggregationScanOperator.java (hunk 4 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/SeriesAggregationScanOperator.java:60]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java (hunk 2 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:101]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java (hunk 3 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:112]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java (hunk 1 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:90]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java [hunk 5]
```diff
@@ -57,6 +59,60 @@ public AlignedSeriesScanUtil(
...
+      throw new IOException("Can only get statistics of time column from alignedChunkMetaData");
+    }
+    return ((AlignedTimeSeriesMetadata) firstTimeSeriesMetadata).getTimeStatistics();
+  }
+
+  @Override
+  protected Statistics currentChunkStatistics(int index) throws IOException {
+    if (!(firstChunkMetadata instanceof AlignedChunkMetadata)) {
+      throw new IOException("Can only get statistics by index from alignedChunkMetaData");
+    }
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:82]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java [hunk 6]
```diff
@@ -57,6 +59,60 @@ public AlignedSeriesScanUtil(
...
+      throw new IOException("Can only get statistics by index from alignedTimeSeriesMetaData");
+    }
+    return ((AlignedTimeSeriesMetadata) firstTimeSeriesMetadata).getStatistics(index);
+  }
+
+  @Override
+  protected Statistics currentFileTimeStatistics() throws IOException {
+    if (!(firstTimeSeriesMetadata instanceof AlignedTimeSeriesMetadata)) {
+      throw new IOException("Can only get statistics of time column from alignedChunkMetaData");
+    }
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:74]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java [hunk 7]
```diff
@@ -57,6 +59,60 @@ public AlignedSeriesScanUtil(
             .getSchemaList().stream().map(IMeasurementSchema::getType).collect(Collectors.toList());
   }
 
+  @Override
+  protected Statistics currentFileStatistics(int index) throws IOException {
+    if (!(firstTimeSeriesMetadata instanceof AlignedTimeSeriesMetadata)) {
+      throw new IOException("Can only get statistics by index from alignedTimeSeriesMetaData");
+    }
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesScanUtil.java:66]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Aggregator.java [hunk 8]
```diff
@@ -108,16 +111,12 @@ public void outputResult(ColumnBuilder[] columnBuilder) {
...
-  }
-
-  /** Used for AlignedSeriesAggregateScanOperator. */
+  /** Used for SeriesAggregateScanOperator. */
   public void processStatistics(Statistics[] statistics) {
     for (InputLocation[] inputLocations : inputLocationList) {
       checkArgument(
           inputLocations[0].getTsBlockIndex() == 0,
-          "AlignedSeriesAggregateScanOperator can only process one tsBlock input.");
+          "SeriesAggregateScanOperator can only process one tsBlock input.");
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Aggregator.java:119]: fixed

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesAggregationScanOperator.java [hunk 9]
```diff
@@ -77,10 +40,8 @@ public AlignedSeriesAggregationScanOperator(
       Filter timeFilter,
       boolean ascending,
       GroupByTimeParameter groupByTimeParameter) {
-    this.sourceId = sourceId;
-    this.operatorContext = context;
-    this.ascending = ascending;
-    this.alignedSeriesScanUtil =
+    super(sourceId, context, aggregators, ascending, groupByTimeParameter);
+    this.seriesScanUtil =
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/source/AlignedSeriesAggregationScanOperator.java:44]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+  public RuleBasedTimeSliceAllocator() {
+    this.operatorToWeightMap = new HashMap<>();
+    this.totalWeight = 0;
+  }
+
+  public void recordExecutionWeight(OperatorContext operatorContext, int weight) {
+    if (operatorToWeightMap.containsKey(operatorContext)) {
+      throw new IllegalArgumentException();
+    }
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java:45]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    operatorToWeightMap.put(operatorContext, weight);
+    totalWeight += weight;
+  }
+
+  public int getWeight(OperatorContext operatorContext) {
+    if (!operatorToWeightMap.containsKey(operatorContext)) {
+      throw new IllegalArgumentException();
+    }
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java:53]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+  public void recordExecutionWeight(OperatorContext operatorContext, int weight) {
+    if (operatorToWeightMap.containsKey(operatorContext)) {
+      throw new IllegalArgumentException();
+    }
+    operatorToWeightMap.put(operatorContext, weight);
+    totalWeight += weight;
+  }
+
+  public int getWeight(OperatorContext operatorContext) {
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java:50]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    if (!operatorToWeightMap.containsKey(operatorContext)) {
+      throw new IllegalArgumentException();
+    }
+    return operatorToWeightMap.get(operatorContext);
+  }
+
+  @Override
+  public Duration getMaxRunTime(OperatorContext operatorContext) {
+    return new Duration(
+        (double) DriverTaskThread.EXECUTION_TIME_SLICE.roundTo(TimeUnit.MILLISECONDS)
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/timer/RuleBasedTimeSliceAllocator.java:60]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java [hunk 14]
```diff
@@ -218,6 +218,13 @@ public DataDriver plan(
 
     Operator root = plan.accept(new Visitor(), context);
 
+    ITimeSliceAllocator timeSliceAllocator = context.getTimeSliceAllocator();
+    context
+        .getAggregationOperatorList()
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java:223]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/AggregationUtil.java [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+          continue;
+        }
+
+        lastReadRowIndex = Math.max(lastReadRowIndex, aggregator.processTsBlock(inputTsBlock));
+      }
+      if (lastReadRowIndex >= inputTsBlock.getPositionCount()) {
+        inputTsBlock = null;
+      } else {
+        inputTsBlock = inputTsBlock.subTsBlock(lastReadRowIndex);
+      }
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/AggregationUtil.java:103]: fixed

File: tsfile/src/main/java/org/apache/iotdb/tsfile/common/conf/TSFileConfig.java [hunk 16]
```diff
@@ -150,6 +150,12 @@ public class TSFileConfig implements Serializable {
   /** The amount of data iterate each time */
   private int batchSize = 1000;
 
+  /** Maximum capacity of a TsBlock */
+  private int maxTsBlockSizeInBytes = 1024 * 1024;
+
+  /** Maximum number of lines in a single TsBlock */
+  private int maxTsBlockLineNumber = 1000;
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/common/conf/TSFileConfig.java:157]: Uh……the parameters in IoTDBConfig are not accessible in the TsBlock.

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Aggregator.java [hunk 17]
```diff
@@ -141,6 +140,7 @@ public boolean hasFinalResult() {
   }
 
   public void updateTimeRange(TimeRange curTimeRange) {
+    reset();
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Aggregator.java:140]: Previously, we called reset() at the same time as updateTimeRange(), so I put reset() in updateTimeRange(). Actually, reset Aggregator is also required when updating time range.

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**